<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.31">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>DataSHIELD Workshop: Data Science Without Borders – DataSHIELD APHRC Workshop</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./images/isglobal_logo.jpg" rel="icon" type="image/jpeg">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-b08616ca30bc6986c6ffe4cecfeccd70.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-d2fec8d93cc0814413f782f2e7a38223.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<style>html{ scroll-behavior: smooth; }</style>
<link href="https://fonts.googleapis.com/css2?family=Inter:wght@300;400;500;600;700&amp;display=swap" rel="stylesheet">


<link rel="stylesheet" href="styles.css">
<link rel="stylesheet" href="navbar-fix.css">
</head>

<body class="nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> <i class="bi bi-house-fill" role="img">
</i> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./getting-started.html"> <i class="bi bi-play-circle-fill" role="img">
</i> 
<span class="menu-text">Getting Started</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-workshops" role="link" data-bs-toggle="dropdown" aria-expanded="false">
      <i class="bi bi-mortarboard-fill" role="img">
</i> 
 <span class="menu-text">Workshops</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-workshops">    
        <li>
    <a class="dropdown-item" href="./datashield-presentation.html"><i class="bi bi-file-earmark-slides" role="img">
</i> 
 <span class="dropdown-text">DataSHIELD Presentation</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./DataSHIELD_workshop_APHRC_tables.html"><i class="bi bi-table" role="img">
</i> 
 <span class="dropdown-text">DataSHIELD (Tables)</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./DataSHIELD_workshop_APHRC_resources.html"><i class="bi bi-link-45deg" role="img">
</i> 
 <span class="dropdown-text">DataSHIELD (Resources)</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./omop-presentation.html"><i class="bi bi-file-earmark-slides" role="img">
</i> 
 <span class="dropdown-text">OMOP CDM Presentation</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./dsomop.html"><i class="bi bi-database-fill" role="img">
</i> 
 <span class="dropdown-text">dsOMOP Workshop</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="./schedule.html"> <i class="bi bi-calendar3" role="img">
</i> 
<span class="menu-text">Schedule</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./resources.html"> <i class="bi bi-book-fill" role="img">
</i> 
<span class="menu-text">Resources</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./contact.html"> <i class="bi bi-envelope-fill" role="img">
</i> 
<span class="menu-text">Contact</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/isglobal-brge"> <i class="bi bi-github" role="img" aria-label="GitHub">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://www.linkedin.com/company/barcelona-institute-for-global-health-isglobal-"> <i class="bi bi-linkedin" role="img" aria-label="LinkedIn">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://www.isglobal.org"> <i class="bi bi-globe" role="img" aria-label="ISGlobal Website">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-full page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#information-for-newcomers-to-rstudio-notebooks" id="toc-information-for-newcomers-to-rstudio-notebooks" class="nav-link active" data-scroll-target="#information-for-newcomers-to-rstudio-notebooks">INFORMATION FOR NEWCOMERS TO RSTUDIO NOTEBOOKS:</a></li>
  <li><a href="#datashield-for-analyzing-simulated-d-data" id="toc-datashield-for-analyzing-simulated-d-data" class="nav-link" data-scroll-target="#datashield-for-analyzing-simulated-d-data">DataSHIELD for analyzing simulated D data</a>
  <ul class="collapse">
  <li><a href="#installing-datashield" id="toc-installing-datashield" class="nav-link" data-scroll-target="#installing-datashield">Installing DataSHIELD</a></li>
  <li><a href="#logging-in-and-assigning-data" id="toc-logging-in-and-assigning-data" class="nav-link" data-scroll-target="#logging-in-and-assigning-data">Logging in and assigning data</a></li>
  <li><a href="#describing-data-aggregate-type-functions" id="toc-describing-data-aggregate-type-functions" class="nav-link" data-scroll-target="#describing-data-aggregate-type-functions">Describing data (‘aggregate-type functions’)</a>
  <ul class="collapse">
  <li><a href="#were-going-to-be-focus-on-hdl" id="toc-were-going-to-be-focus-on-hdl" class="nav-link" data-scroll-target="#were-going-to-be-focus-on-hdl">We’re going to be focus on HDL</a></li>
  <li><a href="#using-dshelper-to-retrieve-statistics-in-a-neater-format." id="toc-using-dshelper-to-retrieve-statistics-in-a-neater-format." class="nav-link" data-scroll-target="#using-dshelper-to-retrieve-statistics-in-a-neater-format.">Using dsHelper to retrieve statistics in a neater format.</a></li>
  </ul></li>
  <li><a href="#manipulating-data-assign-type-functions" id="toc-manipulating-data-assign-type-functions" class="nav-link" data-scroll-target="#manipulating-data-assign-type-functions">Manipulating data (‘assign-type’ functions)</a></li>
  <li><a href="#sub-setting-data" id="toc-sub-setting-data" class="nav-link" data-scroll-target="#sub-setting-data">Sub-setting data</a>
  <ul class="collapse">
  <li><a href="#sub-setting-to-remove-nas" id="toc-sub-setting-to-remove-nas" class="nav-link" data-scroll-target="#sub-setting-to-remove-nas">Sub-setting to remove NAs</a></li>
  <li><a href="#sub-setting-by-inequality" id="toc-sub-setting-by-inequality" class="nav-link" data-scroll-target="#sub-setting-by-inequality">Sub-setting by inequality</a></li>
  <li><a href="#sub-setting-by-multiple-conditions" id="toc-sub-setting-by-multiple-conditions" class="nav-link" data-scroll-target="#sub-setting-by-multiple-conditions">Sub-setting by multiple conditions</a></li>
  </ul></li>
  <li><a href="#data-manipulation-with-dshelper" id="toc-data-manipulation-with-dshelper" class="nav-link" data-scroll-target="#data-manipulation-with-dshelper">Data manipulation with dsHelper</a>
  <ul class="collapse">
  <li><a href="#create-a-subset-of-columns-by-a-vector-of-column-names" id="toc-create-a-subset-of-columns-by-a-vector-of-column-names" class="nav-link" data-scroll-target="#create-a-subset-of-columns-by-a-vector-of-column-names">Create a subset of columns by a vector of column names</a></li>
  <li><a href="#rename-variables" id="toc-rename-variables" class="nav-link" data-scroll-target="#rename-variables">Rename variables</a></li>
  </ul></li>
  <li><a href="#graphs" id="toc-graphs" class="nav-link" data-scroll-target="#graphs">Graphs</a>
  <ul class="collapse">
  <li><a href="#histograms" id="toc-histograms" class="nav-link" data-scroll-target="#histograms">Histograms</a></li>
  <li><a href="#scatterplots-of-two-numerical-variables" id="toc-scatterplots-of-two-numerical-variables" class="nav-link" data-scroll-target="#scatterplots-of-two-numerical-variables">Scatterplots of two numerical variables</a></li>
  </ul></li>
  <li><a href="#analysis" id="toc-analysis" class="nav-link" data-scroll-target="#analysis">Analysis</a>
  <ul class="collapse">
  <li><a href="#simple-linear-regression" id="toc-simple-linear-regression" class="nav-link" data-scroll-target="#simple-linear-regression">Simple Linear Regression</a></li>
  <li><a href="#creating-forest-plots" id="toc-creating-forest-plots" class="nav-link" data-scroll-target="#creating-forest-plots">Creating forest plots</a></li>
  <li><a href="#modelling-multiple-variables-and-interactions" id="toc-modelling-multiple-variables-and-interactions" class="nav-link" data-scroll-target="#modelling-multiple-variables-and-interactions">Modelling multiple variables and interactions</a></li>
  </ul></li>
  <li><a href="#at-the-end-of-your-rstudio-server-analysis" id="toc-at-the-end-of-your-rstudio-server-analysis" class="nav-link" data-scroll-target="#at-the-end-of-your-rstudio-server-analysis">At the end of your RStudio Server analysis:</a></li>
  </ul></li>
  <li><a href="#exercise" id="toc-exercise" class="nav-link" data-scroll-target="#exercise">Exercise</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content column-page-left" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">DataSHIELD Workshop: Data Science Without Borders</h1>
</div>



<div class="quarto-title-meta column-page-left">

    
  
    
  </div>
  


</header>


<p><a href="#" onclick="event.preventDefault(); fetch('https://raw.githubusercontent.com/isglobal-brge/workshop_DSWB/main/DataSHIELD_workshop_APHRC_tables.Rmd').then(response => response.blob()).then(blob => { const url = window.URL.createObjectURL(blob); const a = document.createElement('a'); a.style.display = 'none'; a.href = url; a.download = 'DataSHIELD_workshop_APHRC_tables.Rmd'; document.body.appendChild(a); a.click(); window.URL.revokeObjectURL(url); }); return false;" class="btn btn-primary btn-sm"> <i class="bi bi-download"></i> Download this R Markdown file </a></p>
<section id="information-for-newcomers-to-rstudio-notebooks" class="level3">
<h3 data-anchor-id="information-for-newcomers-to-rstudio-notebooks">INFORMATION FOR NEWCOMERS TO RSTUDIO NOTEBOOKS:</h3>
<p>This is an <a href="http://rmarkdown.rstudio.com">R Markdown</a> Notebook. When you execute code within the notebook, the results appear beneath the code.</p>
<p>Try executing this chunk by clicking the <em>Run</em> button within the chunk or by placing your cursor inside it and pressing <em>Ctrl+Shift+Enter</em>.</p>
</section>
<section id="datashield-for-analyzing-simulated-d-data" class="level1">
<h1>DataSHIELD for analyzing simulated D data</h1>
<p>The plan for this workshop is as follows:</p>
<ul>
<li>Installing DataSHIELD</li>
<li>Logging in and assigning data</li>
<li>Describing data</li>
<li>Manipulating data</li>
<li>Subsetting data</li>
<li>Data manipulation with dsHelper</li>
<li>Making graphs</li>
<li>Performing regression analysis</li>
</ul>
<section id="installing-datashield" class="level2">
<h2 data-anchor-id="installing-datashield">Installing DataSHIELD</h2>
<p>Firstly: check whether we have the right R packages installed to run DataSHIELD: using the very helpful devtools package (which has already been installed for us by Stuart!), we’ll use the “Session info” command:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"devtools"</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(devtools)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">session_info</span>()</span></code></pre></div>
</div>
<p>We are missing some of the necessary packages: “DSI”, “DSOpal” and “dsBaseClient”.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">'DSI'</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">'DSOpal'</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"datashield/dsBaseClient"</span>, <span class="at">force =</span> <span class="cn">TRUE</span>)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"timcadman/ds-helper"</span>)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"metafor"</span>)</span></code></pre></div>
</div>
<p>Remember to load them into this R session using “library()” command:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># libraries to connect with Opal databases</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DSI)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DSOpal)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="co"># DataSHIELD client libraries</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dsBaseClient)</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dsHelper)</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Library to perform meta-analyses</span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(metafor)</span></code></pre></div>
</div>
<p>Check that they have now been added:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">session_info</span>()</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>─ Session info ───────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.2 (2024-10-31)
 os       macOS Sequoia 15.4.1
 system   aarch64, darwin20
 ui       X11
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       Africa/Nairobi
 date     2025-08-18
 pandoc   3.6.3 @ /opt/homebrew/bin/ (via rmarkdown)
 quarto   1.7.31 @ /usr/local/bin/quarto

─ Packages ───────────────────────────────────────────────────────────────────
 package      * version    date (UTC) lib source
 arrow          20.0.0.2   2025-05-26 [1] CRAN (R 4.4.1)
 assertthat     0.2.1      2019-03-21 [1] CRAN (R 4.4.1)
 backports      1.5.0      2024-05-23 [1] CRAN (R 4.4.1)
 bit            4.6.0      2025-03-06 [1] CRAN (R 4.4.1)
 bit64          4.6.0-1    2025-01-16 [1] CRAN (R 4.4.1)
 boot           1.3-31     2024-08-28 [1] CRAN (R 4.4.2)
 cachem         1.1.0      2024-05-16 [1] CRAN (R 4.4.1)
 checkmate      2.3.2      2024-07-29 [1] CRAN (R 4.4.0)
 cli            3.6.5      2025-04-23 [1] CRAN (R 4.4.1)
 crayon         1.5.3      2024-06-20 [1] CRAN (R 4.4.1)
 data.table     1.17.8     2025-07-10 [1] CRAN (R 4.4.1)
 devtools       2.4.5      2022-10-11 [1] CRAN (R 4.4.0)
 digest         0.6.37     2024-08-19 [1] CRAN (R 4.4.1)
 dplyr          1.1.4      2023-11-17 [1] CRAN (R 4.4.0)
 dsBaseClient * 6.3.0      2025-07-03 [1] local
 dsHelper     * 1.7.1      2025-07-03 [1] Github (timcadman/ds-helper@1b2348d)
 DSI          * 1.7.1      2024-11-03 [1] CRAN (R 4.4.1)
 DSOpal       * 1.4.0      2022-10-06 [1] CRAN (R 4.4.0)
 ellipsis       0.3.2      2021-04-29 [1] CRAN (R 4.4.1)
 evaluate       1.0.3      2025-01-10 [1] CRAN (R 4.4.1)
 fastmap        1.2.0      2024-05-15 [1] CRAN (R 4.4.1)
 forcats        1.0.0      2023-01-29 [1] CRAN (R 4.4.0)
 fs             1.6.5      2024-10-30 [1] CRAN (R 4.4.1)
 generics       0.1.4      2025-05-09 [1] CRAN (R 4.4.1)
 glue           1.8.0      2024-09-30 [1] CRAN (R 4.4.1)
 haven          2.5.4      2023-11-30 [1] CRAN (R 4.4.0)
 hms            1.1.3      2023-03-21 [1] CRAN (R 4.4.0)
 htmltools      0.5.8.1    2024-04-04 [1] CRAN (R 4.4.1)
 htmlwidgets    1.6.4      2023-12-06 [1] CRAN (R 4.4.0)
 httpuv         1.6.15     2024-03-26 [1] CRAN (R 4.4.0)
 httr         * 1.4.7      2023-08-15 [1] CRAN (R 4.4.0)
 jsonlite       2.0.0      2025-03-27 [1] CRAN (R 4.4.1)
 knitr          1.49       2024-11-08 [1] CRAN (R 4.4.1)
 labelled       2.14.0     2025-01-08 [1] CRAN (R 4.4.1)
 later          1.4.1      2024-11-27 [1] CRAN (R 4.4.1)
 lattice        0.22-6     2024-03-20 [1] CRAN (R 4.4.2)
 lifecycle      1.0.4      2023-11-07 [1] CRAN (R 4.4.1)
 lme4           1.1-37     2025-03-26 [1] CRAN (R 4.4.1)
 magrittr       2.0.3      2022-03-30 [1] CRAN (R 4.4.1)
 MASS           7.3-61     2024-06-13 [1] CRAN (R 4.4.2)
 mathjaxr       1.8-0      2025-04-30 [1] CRAN (R 4.4.1)
 Matrix       * 1.7-1      2024-10-18 [1] CRAN (R 4.4.2)
 memoise        2.0.1      2021-11-26 [1] CRAN (R 4.4.0)
 metadat      * 1.4-0      2025-02-04 [1] CRAN (R 4.4.1)
 metafor      * 4.8-0      2025-01-28 [1] CRAN (R 4.4.1)
 mime           0.13       2025-03-17 [1] CRAN (R 4.4.1)
 miniUI         0.1.1.1    2018-05-18 [1] CRAN (R 4.4.0)
 minqa          1.2.8      2024-08-17 [1] CRAN (R 4.4.1)
 nlme           3.1-166    2024-08-14 [1] CRAN (R 4.4.2)
 nloptr         2.2.1      2025-03-17 [1] CRAN (R 4.4.1)
 numDeriv     * 2016.8-1.1 2019-06-06 [1] CRAN (R 4.4.1)
 opalr        * 3.4.2      2024-09-18 [1] CRAN (R 4.4.1)
 pillar         1.10.2     2025-04-05 [1] CRAN (R 4.4.1)
 pkgbuild       1.4.6      2025-01-16 [1] CRAN (R 4.4.1)
 pkgconfig      2.0.3      2019-09-22 [1] CRAN (R 4.4.1)
 pkgload        1.4.0      2024-06-28 [1] CRAN (R 4.4.0)
 prettyunits    1.2.0      2023-09-24 [1] CRAN (R 4.4.1)
 profvis        0.4.0      2024-09-20 [1] CRAN (R 4.4.1)
 progress     * 1.2.3      2023-12-06 [1] CRAN (R 4.4.0)
 promises       1.3.2      2024-11-28 [1] CRAN (R 4.4.1)
 purrr          1.0.4      2025-02-05 [1] CRAN (R 4.4.1)
 R6           * 2.6.1      2025-02-15 [1] CRAN (R 4.4.1)
 rbibutils      2.3        2024-10-04 [1] CRAN (R 4.4.1)
 Rcpp           1.0.14     2025-01-12 [1] CRAN (R 4.4.1)
 Rdpack         2.6.4      2025-04-09 [1] CRAN (R 4.4.1)
 readr          2.1.5      2024-01-10 [1] CRAN (R 4.4.0)
 reformulas     0.4.1      2025-04-30 [1] CRAN (R 4.4.1)
 remotes        2.5.0      2024-03-17 [1] CRAN (R 4.4.1)
 rlang          1.1.6      2025-04-11 [1] CRAN (R 4.4.1)
 rmarkdown      2.29       2024-11-04 [1] CRAN (R 4.4.1)
 sessioninfo    1.2.3      2025-02-05 [1] CRAN (R 4.4.1)
 shiny          1.10.0     2024-12-14 [1] CRAN (R 4.4.1)
 stringi        1.8.7      2025-03-27 [1] CRAN (R 4.4.1)
 stringr        1.5.1      2023-11-14 [1] CRAN (R 4.4.0)
 tibble         3.3.0      2025-06-08 [1] CRAN (R 4.4.1)
 tidyr          1.3.1      2024-01-24 [1] CRAN (R 4.4.1)
 tidyselect     1.2.1      2024-03-11 [1] CRAN (R 4.4.0)
 tzdb           0.5.0      2025-03-15 [1] CRAN (R 4.4.1)
 urlchecker     1.0.1      2021-11-30 [1] CRAN (R 4.4.1)
 usethis        3.1.0      2024-11-26 [1] CRAN (R 4.4.1)
 vctrs          0.6.5      2023-12-01 [1] CRAN (R 4.4.0)
 xfun           0.50       2025-01-07 [1] CRAN (R 4.4.1)
 xtable         1.8-4      2019-04-21 [1] CRAN (R 4.4.1)
 yaml           2.3.10     2024-07-26 [1] CRAN (R 4.4.1)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 * ── Packages attached to the search path.

──────────────────────────────────────────────────────────────────────────────</code></pre>
</div>
</div>
</section>
<section id="logging-in-and-assigning-data" class="level2">
<h2 data-anchor-id="logging-in-and-assigning-data">Logging in and assigning data</h2>
<p>The login script has to be customized to fit the data you are trying to connect to.</p>
<p>The “builder &lt;-” and “builder$append” functions are standard features.</p>
<p>For this demonstration we are connecting to simulated data- but if it was data of real people, it would be very important for us not to be able to see individual patients’ information.</p>
<p>For this workshop, we’ll imagine that the data is hosted in a two Opal repositories located in France and Spain. The tables are in a project called CNSIM and the tables are named CNSIM1 and CNSIM2 respectively. Data correspond to two simulated datasets with different numbers of observations of 11 harmonized variables. They contain synthetic data based on a model derived from the participants of the 1958 Birth Cohort, as part of an obesity methodological development project. This dataset does contain some NA values. The available variables are:</p>
<div class="cell">
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<colgroup>
<col style="width: 9%">
<col style="width: 38%">
<col style="width: 4%">
<col style="width: 48%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Variable</th>
<th style="text-align: left;">Description</th>
<th style="text-align: left;">Type</th>
<th style="text-align: left;">Note</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">LAB_TSC</td>
<td style="text-align: left;">Total Serum Cholesterol</td>
<td style="text-align: left;">numeric</td>
<td style="text-align: left;">mmol/L</td>
</tr>
<tr class="even">
<td style="text-align: left;">LAB_TRIG</td>
<td style="text-align: left;">Triglycerides</td>
<td style="text-align: left;">numeric</td>
<td style="text-align: left;">mmol/L</td>
</tr>
<tr class="odd">
<td style="text-align: left;">LAB_HDL</td>
<td style="text-align: left;">HDL Cholesterol</td>
<td style="text-align: left;">numeric</td>
<td style="text-align: left;">mmol/L</td>
</tr>
<tr class="even">
<td style="text-align: left;">LAB_GLUC_ADJUSTED</td>
<td style="text-align: left;">Non-Fasting Glucose</td>
<td style="text-align: left;">numeric</td>
<td style="text-align: left;">mmol/L</td>
</tr>
<tr class="odd">
<td style="text-align: left;">PM_BMI_CONTINUOUS</td>
<td style="text-align: left;">Body Mass Index (continuous)</td>
<td style="text-align: left;">numeric</td>
<td style="text-align: left;">kg/m2</td>
</tr>
<tr class="even">
<td style="text-align: left;">DIS_CVA</td>
<td style="text-align: left;">History of Stroke</td>
<td style="text-align: left;">factor</td>
<td style="text-align: left;">0 = Never had stroke; 1 = Has had stroke</td>
</tr>
<tr class="odd">
<td style="text-align: left;">MEDI_LPD</td>
<td style="text-align: left;">Current Use of Lipid Lowering Medication (from categorical assessment item)</td>
<td style="text-align: left;">factor</td>
<td style="text-align: left;">0 = Not currently using lipid lowering medication; 1 = Currently using lipid lowering medication</td>
</tr>
<tr class="even">
<td style="text-align: left;">DIS_DIAB</td>
<td style="text-align: left;">History of Diabetes</td>
<td style="text-align: left;">factor</td>
<td style="text-align: left;">0 = Never had diabetes; 1 = Has had diabetes</td>
</tr>
<tr class="odd">
<td style="text-align: left;">DIS_AMI</td>
<td style="text-align: left;">History of Myocardial Infarction</td>
<td style="text-align: left;">factor</td>
<td style="text-align: left;">0 = Never had myocardial infarction; 1 = Has had myocardial infarction</td>
</tr>
<tr class="even">
<td style="text-align: left;">GENDER</td>
<td style="text-align: left;">Gender</td>
<td style="text-align: left;">factor</td>
<td style="text-align: left;">0 = Female, 1 = Male</td>
</tr>
<tr class="odd">
<td style="text-align: left;">PM_BMI_CATEGORICAL</td>
<td style="text-align: left;">Body Mass Index (categorical)</td>
<td style="text-align: left;">factor</td>
<td style="text-align: left;">1 = Less than 25 kg/m2; 2 = 25 to 30 kg/m2; 3 = Over 30 kg/m2</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Now, using DataSHIELD, we have to connect to the servers which contains the data in the Opal databases. The below code creates a local R object with the login details for each study:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>builder <span class="ot">&lt;-</span> DSI<span class="sc">::</span><span class="fu">newDSLoginBuilder</span>()</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Server 1: France </span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>builder<span class="sc">$</span><span class="fu">append</span>(</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">server =</span> <span class="st">'France'</span>, </span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">url =</span> <span class="st">"https://opal-demo.obiba.org"</span>,</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">user =</span> <span class="st">"administrator"</span>, </span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">password =</span> <span class="st">"password"</span>,</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">table =</span> <span class="st">"CNSIM.CNSIM1"</span>,</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">profile =</span> <span class="st">"default"</span></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Server 2: Spain (ISGlobal)</span></span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>builder<span class="sc">$</span><span class="fu">append</span>(</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">server =</span> <span class="st">'ISGlobal'</span>, </span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">url =</span> <span class="st">"https://opal.isglobal.org/repo"</span>,</span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a>  <span class="at">user =</span> <span class="st">"invited"</span>, </span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a>  <span class="at">password =</span> <span class="st">"12345678Aa@"</span>,</span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a>  <span class="at">table =</span> <span class="st">"CNSIM.CNSIM2"</span>,</span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a>  <span class="at">profile =</span> <span class="st">"rock-inma"</span></span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
</div>
<p>NOTE: The <code>profile</code> argument is set up by the data owners and controls both, the datasets or projects that are available for this user, as well as the different DataSHIELD packages.</p>
<p>Now we need to connect, referring to the login information in the data frame we have just created:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>logindata <span class="ot">&lt;-</span> builder<span class="sc">$</span><span class="fu">build</span>()</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>conns <span class="ot">&lt;-</span> <span class="fu">datashield.login</span>(<span class="at">logins =</span> logindata, <span class="at">assign =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error: [Client error: (401) Unauthorized]</code></pre>
</div>
</div>
<p>The ‘assign’ argument can be set to either ‘TRUE’ or ‘FALSE’. If set to ‘TRUE’, all the available variables within that table will be assigned to a serverside data frame and available to access. If you only need a small subset of available variables it can be preferable to set this to ‘FALSE’ and later use the function ‘datashield.assign’ to separately assign only the variables you need. The output of this box has useful progress bars which show the progress of connecting to studies, one by one.</p>
<p>We can see the serverside has object called <code>D</code> which correspond to the dataset <code>CNSIM</code> of each study by running:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.ls</span>()</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
</div>
<p>which is a data.frame called <code>D</code> (this name was set using the ‘symbol’ argument in datashield.login above). We can check that this is a data.frame by typing</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.class</span>(<span class="at">x =</span> <span class="st">"D"</span>, <span class="at">datasources =</span> conns)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error: object 'conns' not found</code></pre>
</div>
</div>
<p>NOTE: writting <code>datasources = conns</code> is not required. This is just to emphasize that if you have several connections you need to specify which one is yours. By default, it missing it looks for your local environment. Also, the argument <code>x = "D"</code> can be simply written by <code>"D"</code>.</p>
<p>and the data.frame has the variables we have previously described</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.colnames</span>(<span class="st">"D"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
</div>
<section id="section" class="level80">
<p class="heading"></p>
</section>
</section>
<section id="describing-data-aggregate-type-functions" class="level2">
<h2 data-anchor-id="describing-data-aggregate-type-functions">Describing data (‘aggregate-type functions’)</h2>
<p>There are many data exploration functions already implemented into DataSHIELD: let’s check it out at the wiki <a href="https://data2knowledge.atlassian.net/wiki/spaces/DSDEV/pages/2733244417/Version+6.2.0">https://data2knowledge.atlassian.net/wiki/spaces/DSDEV/pages/2733244417/Version+6.2.0</a></p>
<p>Scroll down to “Data structure queries”. Let’s try out a few of these:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.dim</span>(<span class="at">x=</span><span class="st">"D"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
</div>
<p>What it is <em>mandatory</em> is to write the name of the data.frame with ““.</p>
<section id="were-going-to-be-focus-on-hdl" class="level3">
<h3 data-anchor-id="were-going-to-be-focus-on-hdl">We’re going to be focus on HDL</h3>
<p>This is a measure of HDL Cholesterol (aka the “good cholesterol” level)</p>
<p>Let’s run some summary statistic commands</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.class</span>(<span class="at">x=</span><span class="st">'D$LAB_HDL'</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.length</span>(<span class="at">x=</span><span class="st">'D$LAB_HDL'</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.mean</span>(<span class="at">x=</span><span class="st">'D$LAB_HDL'</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
</div>
<p>What is this other function to obtain the mean? Let’s use the DataSHIELD function help documentation.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>?ds.quantileMean</span></code></pre></div>
</div>
<p>Now, putting into action some of what we’ve learned about the function arguments. NOTE: ‘split’ is in case you have data from different servers and you want to see the statistic one by one.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.quantileMean</span>(<span class="at">x=</span><span class="st">'D$LAB_HDL'</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.quantileMean</span>(<span class="at">x=</span><span class="st">'D$LAB_HDL'</span>, <span class="at">type =</span> <span class="st">"split"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
</div>
<p>Trying to calculate the variance of FEV1:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>?ds.var</span></code></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.var</span>(<span class="at">x =</span> <span class="st">'D$LAB_HDL'</span>, <span class="at">type =</span> <span class="st">"split"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
</div>
<p>Can we store the results calculated from a DataSHIELD analysis in a local R session?</p>
<p>Yes- the output of aggregate functions are always R objects, hence can be stored.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>a<span class="ot">&lt;-</span><span class="fu">ds.var</span>(<span class="at">x =</span> <span class="st">'D$LAB_HDL'</span>, <span class="at">type =</span> <span class="st">"split"</span>)[[<span class="dv">1</span>]]</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>a</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error: object 'a' not found</code></pre>
</div>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>b<span class="ot">&lt;-</span><span class="fu">ds.var</span>(<span class="at">x =</span> <span class="st">'D$LAB_HDL'</span>, <span class="at">type =</span> <span class="st">"split"</span>)[[<span class="dv">1</span>]][[<span class="dv">1</span>,<span class="dv">1</span>]]</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>b</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error: object 'b' not found</code></pre>
</div>
</div>
<p>The square brackets are because we are trying to access an element of a list- which is the R object that DataSHIELD aggregate functions output as.</p>
<p>Factor variables visualize by simply writing</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.table</span>(<span class="st">"D$GENDER"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
</div>
</section>
<section id="using-dshelper-to-retrieve-statistics-in-a-neater-format." class="level3">
<h3 data-anchor-id="using-dshelper-to-retrieve-statistics-in-a-neater-format.">Using dsHelper to retrieve statistics in a neater format.</h3>
<p>As you may have noticed, some operations which are more straightforward in R can be more complicated in datashield. To help with this, the dsHelper package allows you to do some common operations in fewer lines of code. DsHelper is an entirely serverside package - it contains only clientside functions which call DataSHIELD functions serverside.</p>
<p>We have seen datashield has a range of functions to retrieve statistics, but is limited in that (i) you need to use different functions for different statistics, (ii) you can only get stats for one variable at a time. dh.GetStats returns many useful stats in a tibble, and allows you to retrieve stats for multiple variables at a time.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>neat_stats <span class="ot">&lt;-</span> <span class="fu">dh.getStats</span>(</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">df =</span> <span class="st">"D"</span>,</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">vars =</span> <span class="fu">c</span>(<span class="st">"GENDER"</span>, <span class="st">"LAB_TRIG"</span>, <span class="st">"LAB_HDL"</span>, <span class="st">"DIS_CVA"</span>, <span class="st">"DIS_DIAB"</span>))</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>           </span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>neat_stats</span></code></pre></div>
</div>
<section id="section-1" class="level80">
<p class="heading"></p>
</section>
</section>
</section>
<section id="manipulating-data-assign-type-functions" class="level2">
<h2 data-anchor-id="manipulating-data-assign-type-functions">Manipulating data (‘assign-type’ functions)</h2>
<p>Assign-type functions are ones where a calculation is done on the data stored at the server (and results of that calculation are “assigned” to a serverside variable, and saved there), but is NOT transmitted back to the user.</p>
<p>The reason for this is that some calculations could be highly disclosive- and if such data were transmitted to the analyst, with not much effort at all, with an inverse function, the analyst could work out exactly what the raw data are- and thus the data’s privacy is broken!</p>
<p>To demonstrate:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.ls</span>()</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.log</span>(<span class="at">x=</span><span class="st">'D$LAB_HDL'</span>, <span class="at">newobj=</span><span class="st">'HDL_log'</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.ls</span>()</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.mean</span>(<span class="at">x=</span><span class="st">"HDL_log"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.mean</span>(<span class="at">x=</span><span class="st">"D$LAB_HDL"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
</div>
<p>The second “ds.mean” shows that the mean of the logged values are definitely smaller, by about the right amount. The DataSHIELD log function has worked.</p>
<p>There is another DataSHIELD assign function that can be used for data transformations - a square root function. Let’s test again:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.sqrt</span>(<span class="at">x=</span><span class="st">'D$LAB_HDL'</span>, <span class="at">newobj=</span><span class="st">'HDL_sqrt'</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.ls</span>()</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.mean</span>(<span class="at">x=</span><span class="st">"HDL_sqrt"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.mean</span>(<span class="at">x=</span><span class="st">"D$LAB_HDL"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
</div>
<p>These new objects are not attached to a dataframe. Use the help function to find out about the ds.dataFrame function, which can be used to combine objects.</p>
<p>Now join “HDL_sqrt” and “HDL_log” to the dataframe “D”.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.dataFrame</span>(<span class="fu">c</span>(<span class="st">"D"</span>, <span class="st">"HDL_sqrt"</span>, <span class="st">"HDL_log"</span>), <span class="at">newobj =</span> <span class="st">"D"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.colnames</span>(<span class="st">"D"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
</div>
<p><strong>EXERCISE: Using some of the functions above, explore the distribution of the variable “PM_BMI_CATEGORICAL” in dataframe “D”.</strong></p>
<p>Here you see this has returned a list of two tibbles separated into continuous and categorical information. For the categorical variables info is returned on ns, percentages and missingness within each category, whilst for continuous variables info is returned on mean, standard deviation, quantiles and also missingness.</p>
</section>
<section id="sub-setting-data" class="level2">
<h2 data-anchor-id="sub-setting-data">Sub-setting data</h2>
<p>In DataSHIELD there is one function that allows sub-setting of data, ds.dataFrameSubset .</p>
<p>You may wish to use it to:</p>
<p>Subset a column of data by its “Class” Subset a dataframe to remove any “NA”s Subset a numeric column of a dataframe using a Boolean inequalilty</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="co"># first find the column name you wish to refer to</span></span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.colnames</span>(<span class="at">x=</span><span class="st">"D"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="co"># then check which levels you need to apply a boolean operator to:</span></span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.levels</span>(<span class="at">x=</span><span class="st">"D$GENDER"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>?ds.dataFrameSubset</span></code></pre></div>
</div>
<p>Splitting into GENDER groups, assigned to different server-side objects.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.dataFrameSubset</span>(<span class="at">df.name =</span> <span class="st">"D"</span>, <span class="at">V1.name =</span> <span class="st">"D$GENDER"</span>, <span class="at">V2.name =</span> <span class="st">"1"</span>, </span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>                   <span class="at">Boolean.operator =</span> <span class="st">"=="</span>, <span class="at">newobj =</span> <span class="st">"CNSIM.subset.Males"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.dataFrameSubset</span>(<span class="at">df.name =</span> <span class="st">"D"</span>, <span class="at">V1.name =</span> <span class="st">"D$GENDER"</span>, <span class="at">V2.name =</span> <span class="st">"0"</span>, </span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>                   <span class="at">Boolean.operator =</span> <span class="st">"=="</span>, <span class="at">newobj =</span> <span class="st">"CNSIM.subset.Females"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
</div>
<p>Now there are two serverside objects which have split GENDER by class, to which we have assigned the names “CNSIM.subset.Males” and “CNSIM.subset.Females”.</p>
<section id="sub-setting-to-remove-nas" class="level3">
<h3 data-anchor-id="sub-setting-to-remove-nas">Sub-setting to remove NAs</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.completeCases</span>(<span class="at">x1=</span><span class="st">"D"</span>,<span class="at">newobj=</span><span class="st">"D_without_NA"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
</div>
</section>
<section id="sub-setting-by-inequality" class="level3">
<h3 data-anchor-id="sub-setting-by-inequality">Sub-setting by inequality</h3>
<p>Say we wanted to have a subset of patients where BMI values are ≥ 25, and call it subset.BMI.25.plus</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.dataFrameSubset</span>(<span class="at">df.name =</span> <span class="st">"D"</span>,</span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">V1.name =</span> <span class="st">"D$PM_BMI_CONTINUOUS"</span>,</span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">V2.name =</span> <span class="st">"25"</span>,</span>
<span id="cb75-4"><a href="#cb75-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">Boolean.operator =</span> <span class="st">"&gt;="</span>,</span>
<span id="cb75-5"><a href="#cb75-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">newobj =</span> <span class="st">"subset.BMI.25.plus"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
</div>
<p>Checking we have successfully created such an object, using quantiles and histograms:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.quantileMean</span>(<span class="at">x=</span><span class="st">"subset.BMI.25.plus$PM_BMI_CONTINUOUS"</span>, <span class="at">type =</span> <span class="st">"split"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.histogram</span>(<span class="at">x=</span><span class="st">"subset.BMI.25.plus$PM_BMI_CONTINUOUS"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
</div>
</section>
<section id="sub-setting-by-multiple-conditions" class="level3">
<h3 data-anchor-id="sub-setting-by-multiple-conditions">Sub-setting by multiple conditions</h3>
<p>If we want to create a subset based on multiple conditions we can use the ds.Boole function before subsetting. For example, let’s say that we want to create a subset of individuals where BMI values are ≥ 25 and adjusted glucose is lower than 6.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.Boole</span>(</span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">V1 =</span> <span class="st">"D$PM_BMI_CONTINUOUS"</span>,</span>
<span id="cb81-3"><a href="#cb81-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">V2 =</span> <span class="st">"25"</span>,</span>
<span id="cb81-4"><a href="#cb81-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">Boolean.operator =</span> <span class="st">"&gt;="</span>,</span>
<span id="cb81-5"><a href="#cb81-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">numeric.output =</span> <span class="cn">TRUE</span>,</span>
<span id="cb81-6"><a href="#cb81-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">newobj =</span> <span class="st">"BMI.25.plus"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.Boole</span>(</span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">V1 =</span> <span class="st">"D$LAB_GLUC_ADJUSTED"</span>,</span>
<span id="cb83-3"><a href="#cb83-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">V2 =</span> <span class="st">"6"</span>,</span>
<span id="cb83-4"><a href="#cb83-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">Boolean.operator =</span> <span class="st">"&lt;"</span>,</span>
<span id="cb83-5"><a href="#cb83-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">numeric.output =</span> <span class="cn">TRUE</span>,</span>
<span id="cb83-6"><a href="#cb83-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">newobj =</span> <span class="st">"GLUC.6.less"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
</div>
<p>We can then use the ds.make function to make a new categorical variable which combines these groups:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a>?ds.make </span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb85-3"><a href="#cb85-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.make</span>(<span class="at">toAssign =</span> <span class="st">"BMI.25.plus+GLUC.6.less"</span>,</span>
<span id="cb85-4"><a href="#cb85-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">newobj =</span> <span class="st">"BMI.25.plus_GLUC.6.less"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="co"># If BMI &gt;= 25 and glucose &lt; 6, then BMI.25.plus_GLUC.6.less=2</span></span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a><span class="co"># If BMI &gt;= 25 and glucose &gt;= 6, then BMI.25.plus_GLUC.6.less=1</span></span>
<span id="cb87-3"><a href="#cb87-3" aria-hidden="true" tabindex="-1"></a><span class="co"># If BMI &lt; 25 and glucose &lt; 6, then BMI.25.plus_GLUC.6.less=1</span></span>
<span id="cb87-4"><a href="#cb87-4" aria-hidden="true" tabindex="-1"></a><span class="co"># If BMI &lt; 25 and glucose &gt;= 6, then BMI.25.plus_GLUC.6.less=0</span></span>
<span id="cb87-5"><a href="#cb87-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb87-6"><a href="#cb87-6" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.table</span>(<span class="at">rvar=</span> <span class="st">"BMI.25.plus_GLUC.6.less"</span>,</span>
<span id="cb87-7"><a href="#cb87-7" aria-hidden="true" tabindex="-1"></a>         <span class="at">datasources =</span> conns)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error: object 'conns' not found</code></pre>
</div>
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.dataFrame</span>(<span class="at">x=</span><span class="fu">c</span>(<span class="st">"D"</span>, <span class="st">"BMI.25.plus_GLUC.6.less"</span>), <span class="at">newobj =</span> <span class="st">"D2"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.colnames</span>(<span class="st">"D2"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.dataFrameSubset</span>(<span class="at">df.name =</span> <span class="st">"D2"</span>,</span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">V1.name =</span> <span class="st">"D2$BMI.25.plus_GLUC.6.less"</span>,</span>
<span id="cb93-3"><a href="#cb93-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">V2.name =</span> <span class="st">"2"</span>,</span>
<span id="cb93-4"><a href="#cb93-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">Boolean.operator =</span> <span class="st">"=="</span>,</span>
<span id="cb93-5"><a href="#cb93-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">newobj =</span> <span class="st">"subset2"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.dim</span>(<span class="st">"subset2"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
</div>
</section>
</section>
<section id="data-manipulation-with-dshelper" class="level2">
<h2 data-anchor-id="data-manipulation-with-dshelper">Data manipulation with dsHelper</h2>
<p>Again, we can use some dsHelper functions to do data manipulation operations in a more efficient way.</p>
<section id="create-a-subset-of-columns-by-a-vector-of-column-names" class="level3">
<h3 data-anchor-id="create-a-subset-of-columns-by-a-vector-of-column-names">Create a subset of columns by a vector of column names</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dh.dropCols</span>(</span>
<span id="cb97-2"><a href="#cb97-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">df =</span> <span class="st">"D"</span>, </span>
<span id="cb97-3"><a href="#cb97-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">vars =</span> <span class="fu">c</span>(<span class="st">"PM_BMI_CONTINUOUS"</span>, <span class="st">"GENDER"</span>), </span>
<span id="cb97-4"><a href="#cb97-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">type =</span> <span class="st">"keep"</span>,</span>
<span id="cb97-5"><a href="#cb97-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">new_obj =</span> <span class="st">"df_subset"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.colnames</span>(<span class="st">"df_subset"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
</div>
</section>
<section id="rename-variables" class="level3">
<h3 data-anchor-id="rename-variables">Rename variables</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dh.renameVars</span>(</span>
<span id="cb101-2"><a href="#cb101-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">df =</span> <span class="st">"D"</span>, </span>
<span id="cb101-3"><a href="#cb101-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">current_names =</span> <span class="fu">c</span>(<span class="st">"PM_BMI_CONTINUOUS"</span>, <span class="st">"GENDER"</span>),</span>
<span id="cb101-4"><a href="#cb101-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">new_names =</span> <span class="fu">c</span>(<span class="st">"bmi"</span>, <span class="st">"sex"</span>), </span>
<span id="cb101-5"><a href="#cb101-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">new_obj =</span> <span class="st">"df_rename"</span>)</span>
<span id="cb101-6"><a href="#cb101-6" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb101-7"><a href="#cb101-7" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.colnames</span>(<span class="st">"df_rename"</span>)</span></code></pre></div>
</div>
<p>There are many more dsHelper functions designed to make common operations easier in datashield, check out the vignettes at: <a href="https://github.com/timcadman/ds-helper/blob/master/vignettes/ds-helper-main-vignette.Rmd">https://github.com/timcadman/ds-helper/blob/master/vignettes/ds-helper-main-vignette.Rmd</a></p>
<section id="section-2" class="level80">
<p class="heading"></p>
</section>
</section>
</section>
<section id="graphs" class="level2">
<h2 data-anchor-id="graphs">Graphs</h2>
<p>Visualising the data we are studying is extremely important to get a sense of it. While it may seem disclosive at first glance, only such graphs that are definitively non-disclosive have been implemented within the DataSHIELD project.</p>
<section id="histograms" class="level3">
<h3 data-anchor-id="histograms">Histograms</h3>
<p>Firstly, histograms give a good sense of how one variable is distributed. But no individual points are disclosed because values are “binned” into groups of a similar magnitude, disguising what each one actually is. We protect privacy by removing bins with low counts (below specific threshold). If you have a symmetric distribution, you may find some things aren’t observed at the extreme ends.</p>
<p>Let’s create a histogram of the variable we’ve been investigating for much of this study: HDL Cholesterol (“LAB_HDL”).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a>?ds.histogram</span>
<span id="cb102-2"><a href="#cb102-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.histogram</span>(<span class="at">x=</span><span class="st">'D$LAB_HDL'</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
</div>
<p><strong>Use the ds.histogram to explore the distribution of “D$PM_BMI_CONTINUOUS”</strong></p>
</section>
<section id="scatterplots-of-two-numerical-variables" class="level3">
<h3 data-anchor-id="scatterplots-of-two-numerical-variables">Scatterplots of two numerical variables</h3>
<p>When you generate a scatter plot, you can say that the data points that are displayed are not the actual values. The function gives you the choice on how to anonymise: either you anonymise the values by additional random noise; or you take the average of the k nearest neighbours. (for more details on how anonymisation methods are used for the generation of privacy-preserving visualisations you can have a look on the paper https://epjdatascience.springeropen.com/articles/10.1140/epjds/s13688-020-00257-4)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.scatterPlot</span>(<span class="at">x=</span><span class="st">"D$LAB_HDL"</span>, <span class="at">y=</span><span class="st">"D$PM_BMI_CONTINUOUS"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
</div>
<p>Other DataSHIELD graphical functions allow the creation of box plots, heatmap plots and contour plots. Investigate them using their help functions:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a>?ds.heatmapPlot</span>
<span id="cb106-2"><a href="#cb106-2" aria-hidden="true" tabindex="-1"></a>?ds.contourPlot</span>
<span id="cb106-3"><a href="#cb106-3" aria-hidden="true" tabindex="-1"></a>?ds.boxPlot</span></code></pre></div>
</div>
<section id="section-3" class="level80">
<p class="heading"></p>
</section>
</section>
</section>
<section id="analysis" class="level2">
<h2 data-anchor-id="analysis">Analysis</h2>
<section id="simple-linear-regression" class="level3">
<h3 data-anchor-id="simple-linear-regression">Simple Linear Regression</h3>
<p>We want to examine the relationship between BMI and HDL Cholesterol</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.cor</span>(<span class="at">x=</span><span class="st">'D$PM_BMI_CONTINUOUS'</span>, <span class="at">y=</span><span class="st">'D$LAB_HDL'</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
</div>
<p>Regress HDL Cholesterol with BMI using the Individual Partition Data (IPD) approach:</p>
<p>The method for this (ds.glm) is a “pooled analysis”- equivalent to placing the individual-level data from all sources in one warehouse.</p>
<p>Important to note that the link function is by default the canonical link function for each family. So binomial &lt;-&gt; logistic link, poisson &lt;-&gt; log link, gaussian &lt;-&gt; identity link.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.glm</span>(<span class="at">formula =</span> <span class="st">"D$LAB_HDL~D$PM_BMI_CONTINUOUS"</span>, </span>
<span id="cb109-2"><a href="#cb109-2" aria-hidden="true" tabindex="-1"></a>       <span class="at">family=</span><span class="st">"gaussian"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
</div>
<p>Regress HDL Cholesterol with BMI using the Study-Level Meta-Analysis (SLMA) approach:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.glmSLMA</span>(<span class="at">formula =</span> <span class="st">"D$LAB_HDL~D$PM_BMI_CONTINUOUS"</span>, <span class="at">family=</span><span class="st">"gaussian"</span>, </span>
<span id="cb111-2"><a href="#cb111-2" aria-hidden="true" tabindex="-1"></a>           <span class="at">newobj =</span> <span class="st">"workshop.obj"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
</div>
<p>For the SLMA approach we can assign the predicted values at each study:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb113"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb113-1"><a href="#cb113-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.glmPredict</span>(<span class="at">glmname =</span> <span class="st">"workshop.obj"</span>, <span class="at">newobj =</span> <span class="st">"workshop.prediction.obj"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
<div class="sourceCode cell-code" id="cb115"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb115-1"><a href="#cb115-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.length</span>(<span class="st">"workshop.prediction.obj$fit"</span>, <span class="at">datasources=</span>conns)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error: object 'conns' not found</code></pre>
</div>
<div class="sourceCode cell-code" id="cb117"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb117-1"><a href="#cb117-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.length</span>(<span class="st">"D$LAB_HDL"</span>, <span class="at">datasources=</span>conns)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error: object 'conns' not found</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb119"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb119-1"><a href="#cb119-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.cbind</span>(<span class="fu">c</span>(<span class="st">'D$LAB_HDL'</span>, <span class="st">'D$PM_BMI_CONTINUOUS'</span>), <span class="at">newobj=</span><span class="st">'vars'</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
<div class="sourceCode cell-code" id="cb121"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb121-1"><a href="#cb121-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.completeCases</span>(<span class="st">'vars'</span>, <span class="at">newobj=</span><span class="st">'vars.complete'</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
<div class="sourceCode cell-code" id="cb123"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb123-1"><a href="#cb123-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.dim</span>(<span class="st">'vars.complete'</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
</div>
<p>Let’s plot the best linear fit on a scatter plot</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb125"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb125-1"><a href="#cb125-1" aria-hidden="true" tabindex="-1"></a>df1 <span class="ot">&lt;-</span> <span class="fu">ds.scatterPlot</span>(<span class="st">'D$PM_BMI_CONTINUOUS'</span>, <span class="st">"D$LAB_HDL"</span>, <span class="at">return.coords =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
<div class="sourceCode cell-code" id="cb127"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb127-1"><a href="#cb127-1" aria-hidden="true" tabindex="-1"></a>df2 <span class="ot">&lt;-</span> <span class="fu">ds.scatterPlot</span>(<span class="st">'vars.complete$PM_BMI_CONTINUOUS'</span>, <span class="st">"workshop.prediction.obj$fit"</span>, </span>
<span id="cb127-2"><a href="#cb127-2" aria-hidden="true" tabindex="-1"></a>                      <span class="at">return.coords =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
<div class="sourceCode cell-code" id="cb129"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb129-1"><a href="#cb129-1" aria-hidden="true" tabindex="-1"></a><span class="co"># then in native R</span></span>
<span id="cb129-2"><a href="#cb129-2" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">2</span>))</span>
<span id="cb129-3"><a href="#cb129-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">as.data.frame</span>(df1[[<span class="dv">1</span>]][[<span class="dv">1</span>]])<span class="sc">$</span>x,</span>
<span id="cb129-4"><a href="#cb129-4" aria-hidden="true" tabindex="-1"></a>     <span class="fu">as.data.frame</span>(df1[[<span class="dv">1</span>]][[<span class="dv">1</span>]])<span class="sc">$</span>y, <span class="at">xlab=</span><span class="st">'Body Mass Index'</span>, <span class="at">ylab=</span><span class="st">'HDL Cholesterol'</span>, <span class="at">main=</span><span class="st">'Study 1'</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error: object 'df1' not found</code></pre>
</div>
<div class="sourceCode cell-code" id="cb131"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb131-1"><a href="#cb131-1" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(<span class="fu">as.data.frame</span>(df2[[<span class="dv">1</span>]][[<span class="dv">1</span>]])<span class="sc">$</span>x,<span class="fu">as.data.frame</span>(df2[[<span class="dv">1</span>]][[<span class="dv">1</span>]])<span class="sc">$</span>y, <span class="at">col=</span><span class="st">'red'</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error: object 'df2' not found</code></pre>
</div>
<div class="sourceCode cell-code" id="cb133"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb133-1"><a href="#cb133-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">as.data.frame</span>(df1[[<span class="dv">1</span>]][[<span class="dv">2</span>]])<span class="sc">$</span>x,<span class="fu">as.data.frame</span>(df1[[<span class="dv">1</span>]][[<span class="dv">2</span>]])<span class="sc">$</span>y, </span>
<span id="cb133-2"><a href="#cb133-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab=</span><span class="st">'Body Mass Index'</span>, <span class="at">ylab=</span><span class="st">'HDL Cholesterol'</span>, <span class="at">main=</span><span class="st">'Study 2'</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error: object 'df1' not found</code></pre>
</div>
<div class="sourceCode cell-code" id="cb135"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb135-1"><a href="#cb135-1" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(<span class="fu">as.data.frame</span>(df2[[<span class="dv">1</span>]][[<span class="dv">2</span>]])<span class="sc">$</span>x,<span class="fu">as.data.frame</span>(df2[[<span class="dv">1</span>]][[<span class="dv">2</span>]])<span class="sc">$</span>y, <span class="at">col=</span><span class="st">'red'</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error: object 'df2' not found</code></pre>
</div>
</div>
<p>For the SLMA approach we can also create the predicted values and the residuals at each study using the ds.make function:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb137"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb137-1"><a href="#cb137-1" aria-hidden="true" tabindex="-1"></a>glmslma <span class="ot">&lt;-</span> <span class="fu">ds.glmSLMA</span>(<span class="at">formula =</span> <span class="st">"vars.complete$LAB_HDL~vars.complete$PM_BMI_CONTINUOUS"</span>, <span class="at">family=</span><span class="st">"gaussian"</span>, <span class="at">newobj =</span> <span class="st">"workshop.obj"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
<div class="sourceCode cell-code" id="cb139"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb139-1"><a href="#cb139-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.make</span>(<span class="at">toAssign=</span><span class="fu">paste0</span>(<span class="st">"("</span>,glmslma<span class="sc">$</span>SLMA.pooled.ests.matrix[<span class="dv">1</span>,<span class="dv">1</span>],<span class="st">")+("</span>, glmslma<span class="sc">$</span>SLMA.pooled.ests.matrix[<span class="dv">2</span>,<span class="dv">1</span>],<span class="st">"*vars.complete$PM_BMI_CONTINUOUS)"</span>), </span>
<span id="cb139-2"><a href="#cb139-2" aria-hidden="true" tabindex="-1"></a>        <span class="at">newobj =</span> <span class="st">"predicted.values"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
<div class="sourceCode cell-code" id="cb141"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb141-1"><a href="#cb141-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.make</span>(<span class="at">toAssign =</span> <span class="st">"vars.complete$LAB_HDL - predicted.values"</span>, </span>
<span id="cb141-2"><a href="#cb141-2" aria-hidden="true" tabindex="-1"></a>        <span class="at">newobj =</span> <span class="st">"residuals"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
<div class="sourceCode cell-code" id="cb143"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb143-1"><a href="#cb143-1" aria-hidden="true" tabindex="-1"></a><span class="co"># and you can use those to run regression plot diagnostics  </span></span>
<span id="cb143-2"><a href="#cb143-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.scatterPlot</span>(<span class="st">'predicted.values'</span>, <span class="st">"residuals"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
<div class="sourceCode cell-code" id="cb145"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb145-1"><a href="#cb145-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.histogram</span>(<span class="st">"residuals"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
</div>
</section>
<section id="creating-forest-plots" class="level3">
<h3 data-anchor-id="creating-forest-plots">Creating forest plots</h3>
<p>We want to examine the relationship between BMI and diabetes</p>
<p>Examine the distribution of the variable “DIS_DIAB” in all cohorts using ‘ds.table’:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb147"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb147-1"><a href="#cb147-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.table</span>(<span class="st">"D$DIS_DIAB"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
</div>
<p>Check the class of “DIS_DIAB”:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb149"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb149-1"><a href="#cb149-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.class</span>(<span class="st">"D$DIS_DIAB"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
</div>
<p>Examine the association between BMI and diabetes:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb151"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb151-1"><a href="#cb151-1" aria-hidden="true" tabindex="-1"></a>glmSLMA_mod2<span class="ot">&lt;-</span><span class="fu">ds.glmSLMA</span>(<span class="at">formula=</span><span class="st">"D$DIS_DIAB~D$PM_BMI_CONTINUOUS"</span>, <span class="at">family=</span><span class="st">'binomial'</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
</div>
<p>Save effect estimates and standard errors as new objects</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb153"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb153-1"><a href="#cb153-1" aria-hidden="true" tabindex="-1"></a>estimates <span class="ot">&lt;-</span> <span class="fu">c</span>(glmSLMA_mod2<span class="sc">$</span>betamatrix.valid[<span class="dv">2</span>,])</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error: object 'glmSLMA_mod2' not found</code></pre>
</div>
<div class="sourceCode cell-code" id="cb155"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb155-1"><a href="#cb155-1" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">c</span>(glmSLMA_mod2<span class="sc">$</span>sematrix.valid[<span class="dv">2</span>,])</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error: object 'glmSLMA_mod2' not found</code></pre>
</div>
</div>
<p>Meta-analyse the results using rma to obtain study weights:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb157"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb157-1"><a href="#cb157-1" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">rma</span>(estimates, <span class="at">sei=</span>se)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error: Cannot find the object/variable ('estimates') specified for the 'yi' argument.</code></pre>
</div>
</div>
<p>Can produce simple forest plots using output:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb159"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb159-1"><a href="#cb159-1" aria-hidden="true" tabindex="-1"></a><span class="fu">forest</span>(res, <span class="at">atransf=</span>exp)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error: object 'res' not found</code></pre>
</div>
</div>
<p>We can also add more information to forest plots:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb161"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb161-1"><a href="#cb161-1" aria-hidden="true" tabindex="-1"></a>study_names <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"France"</span>, <span class="st">"Spain"</span>)</span>
<span id="cb161-2"><a href="#cb161-2" aria-hidden="true" tabindex="-1"></a>weights <span class="ot">&lt;-</span>  <span class="fu">c</span>(<span class="fu">paste0</span>(<span class="fu">formatC</span>(<span class="fu">weights</span>(res), <span class="at">format=</span><span class="st">"f"</span>, <span class="at">digits=</span><span class="dv">1</span>, <span class="at">width=</span><span class="dv">4</span>), <span class="st">"%"</span>))</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error: object 'res' not found</code></pre>
</div>
<div class="sourceCode cell-code" id="cb163"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb163-1"><a href="#cb163-1" aria-hidden="true" tabindex="-1"></a><span class="fu">forest</span>(res, <span class="at">atransf=</span>exp,</span>
<span id="cb163-2"><a href="#cb163-2" aria-hidden="true" tabindex="-1"></a>       <span class="at">xlab=</span><span class="st">"Crude Odds Ratio"</span>, <span class="at">refline=</span><span class="fu">log</span>(<span class="dv">1</span>), <span class="at">xlim=</span><span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.25</span>,<span class="fl">0.5</span>), </span>
<span id="cb163-3"><a href="#cb163-3" aria-hidden="true" tabindex="-1"></a>       <span class="at">at=</span><span class="fu">log</span>(<span class="fu">c</span>(<span class="fl">0.95</span>, <span class="dv">1</span>, <span class="fl">1.1</span>, <span class="fl">1.2</span>, <span class="fl">1.3</span>)),</span>
<span id="cb163-4"><a href="#cb163-4" aria-hidden="true" tabindex="-1"></a>       <span class="at">slab=</span><span class="fu">cbind</span>(<span class="fu">paste0</span>(study_names, <span class="st">" ("</span>, <span class="fu">paste0</span>(weights, <span class="st">")"</span>))), </span>
<span id="cb163-5"><a href="#cb163-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">mlab=</span><span class="st">"RE model"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error: object 'res' not found</code></pre>
</div>
<div class="sourceCode cell-code" id="cb165"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb165-1"><a href="#cb165-1" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="fl">0.5</span>, <span class="fl">4.5</span>, <span class="at">pos=</span><span class="dv">2</span>, <span class="st">"Odds Ratio [95% CI]"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error in text.default(0.5, 4.5, pos = 2, "Odds Ratio [95% CI]"): plot.new has not been called yet</code></pre>
</div>
<div class="sourceCode cell-code" id="cb167"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb167-1"><a href="#cb167-1" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="sc">-</span><span class="fl">0.25</span>, <span class="fl">4.5</span>, <span class="at">pos=</span><span class="dv">4</span>, <span class="st">"Study (weight)"</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error in text.default(-0.25, 4.5, pos = 4, "Study (weight)"): plot.new has not been called yet</code></pre>
</div>
</div>
</section>
<section id="modelling-multiple-variables-and-interactions" class="level3">
<h3 data-anchor-id="modelling-multiple-variables-and-interactions">Modelling multiple variables and interactions</h3>
<p>Also possible to model multiple explanatory variables and include interactions:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb169"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb169-1"><a href="#cb169-1" aria-hidden="true" tabindex="-1"></a>glm_mod1<span class="ot">&lt;-</span><span class="fu">ds.glm</span>(<span class="at">formula=</span><span class="st">"D$DIS_DIAB~D$PM_BMI_CONTINUOUS+D$LAB_HDL*D$GENDER"</span>, <span class="at">family=</span><span class="st">'binomial'</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
</div>
<p>The “*” between LAB_HDL and GENDER means fit all possible main effects and interactions between the two covariates.</p>
<p>Compare with results of a study-level meta analysis:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb171"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb171-1"><a href="#cb171-1" aria-hidden="true" tabindex="-1"></a>glmSLMA_mod2<span class="ot">&lt;-</span><span class="fu">ds.glmSLMA</span>(<span class="at">formula=</span><span class="st">"D$DIS_DIAB~D$PM_BMI_CONTINUOUS+D$LAB_HDL*D$GENDER"</span>, <span class="at">family=</span><span class="st">'binomial'</span>)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error:  Are you logged in to any server? Please provide a valid DSConnection object! </code></pre>
</div>
</div>
<p>Now compare outputs:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb173"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb173-1"><a href="#cb173-1" aria-hidden="true" tabindex="-1"></a>glm_mod1<span class="sc">$</span>coefficients</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error: object 'glm_mod1' not found</code></pre>
</div>
<div class="sourceCode cell-code" id="cb175"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb175-1"><a href="#cb175-1" aria-hidden="true" tabindex="-1"></a>glmSLMA_mod2<span class="sc">$</span>SLMA.pooled.ests.matrix</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error: object 'glmSLMA_mod2' not found</code></pre>
</div>
</div>
<p>Similar, but differences between the results are accounted for by the different techniques employed.</p>
</section>
</section>
<section id="at-the-end-of-your-rstudio-server-analysis" class="level2">
<h2 data-anchor-id="at-the-end-of-your-rstudio-server-analysis">At the end of your RStudio Server analysis:</h2>
<p>You can save your workspace:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb177"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb177-1"><a href="#cb177-1" aria-hidden="true" tabindex="-1"></a><span class="fu">datashield.workspace_save</span>(<span class="at">conns =</span> conns, <span class="at">ws =</span> <span class="st">"workspace2025"</span>)</span></code></pre></div>
</div>
<p>Don’t forget to log out! Using:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb178"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb178-1"><a href="#cb178-1" aria-hidden="true" tabindex="-1"></a><span class="fu">datashield.logout</span>(conns)</span></code></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error: object 'conns' not found</code></pre>
</div>
</div>
<p>You can restore your workspace, the next time you want to continue with your analysis</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb180"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb180-1"><a href="#cb180-1" aria-hidden="true" tabindex="-1"></a>conns <span class="ot">&lt;-</span> <span class="fu">datashield.login</span>(<span class="at">logins =</span> logindata, </span>
<span id="cb180-2"><a href="#cb180-2" aria-hidden="true" tabindex="-1"></a>                          <span class="at">assign =</span> <span class="cn">TRUE</span>, <span class="at">symbol =</span> <span class="st">"D"</span>)</span>
<span id="cb180-3"><a href="#cb180-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.ls</span>()</span>
<span id="cb180-4"><a href="#cb180-4" aria-hidden="true" tabindex="-1"></a><span class="fu">datashield.logout</span>(conns)</span>
<span id="cb180-5"><a href="#cb180-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb180-6"><a href="#cb180-6" aria-hidden="true" tabindex="-1"></a>conns <span class="ot">&lt;-</span> <span class="fu">datashield.login</span>(<span class="at">logins =</span> logindata, <span class="at">restore =</span> <span class="st">"workspace2025"</span>)</span>
<span id="cb180-7"><a href="#cb180-7" aria-hidden="true" tabindex="-1"></a><span class="fu">ds.ls</span>()</span></code></pre></div>
</div>
<p>Also you can delete unwanted workspaces using the datashield.workspace_rm</p>
<p>In Rstudio Server: DON’T forget to use the orange “quit the current R session” button (top right of browser screen) before closing the tab- otherwise you will experience an error message the next time you try to log in.</p>
</section>
</section>
<section id="exercise" class="level1">
<h1>Exercise</h1>
<p>We have access to 2 datasets corresponding to simulated data from <a href="https://www.ukbiobank.ac.uk/">UKBiobank</a> available through <a href="https://ega-archive.org/datasets/EGAD00001006673">CINECA study</a>. This data reproduce the exact associations found at UKBiobank. The two datasets are accessed in two Opal server</p>
<p>Server 1:<br>
- URL: https://opal-demo.obiba.org/ (user: administrator, pwd: password)<br>
- Project: GWAS<br>
- Table: ega_phenotypes_1</p>
<p>Server 2:<br>
- URL: https://opal.isglobal.org/repo (user: invited, pwd: P@12345678Aa@)<br>
- Project: EGA<br>
- Table: ega_phenotypes_2</p>
<p>Then, load ONE of the three resources in R as data.frame’s using the functions available in the DSI library and answer the next questions using the functions available at dsBaseClient package.</p>
<ul>
<li>Check that your loaded objects are of class data.frame</li>
<li>How many individuals have been diagnosed with diabetes by doctor (variable - diabetes_diagnosed_doctor)?</li>
<li>Obtain the same information stratified by sex (Hint: create a 2x2 table).</li>
<li>Create an histogram of the variable height by combining information across the three different datasets (Hint: type ?ds.histogram to see how to get this plot).</li>
<li>Create a correlation plot between bmi and weight combining data from the three studies (Hint: ?ds.scatterPlot).</li>
<li>Compute the correlation between bmiand weight.</li>
<li>Fit a regression model between cholesterol and weight.</li>
<li>Fit a regression model between diabetes (variable diabetes_diagnosed_doctor) and colesterol level (variable cholesterol). Note: remember that outcome variable (e.g.&nbsp;diabetes) must be a factor variable.</li>
<li>Fit the same model adjusted by bmi. Is still cholesterol associated with diabetes?</li>
<li>Is there any interaction between cholesterol and sex adjusted by bmi?</li>
</ul>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/isglobal-brge\.github\.io\/workshop_APHRC");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
<p>© 2025 <a href="https://www.isglobal.org">ISGlobal - Barcelona Institute for Global Health</a><br> <a href="https://brge.isglobal.org">Bioinformatics Research Group in Epidemiology (BRGE)</a></p>
</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>
<script src="workshop-interactive.js"></script>




<script src="site_libs/quarto-html/zenscroll-min.js"></script>
</body></html>